src/api/cocoa/mod.rs

```diff
@@ -1,28 +1,307 @@
+                                             .autorelease()
+        };
+
+        let path = unsafe {
+            NSString::alloc(nil).init_str(icon_file)
+        };
+        if path == nil {
+            return Err(Error::OsError("Could not create `NSString` out of the passed &str"
+                                             .to_owned()));
+        }
+
+        let nsimage = unsafe { NSImage::initWithContentsOfFile_(NSImage::alloc(nil), path).autorelease() };
+        if nsimage == nil {
+            return Err(Error::OsError("Could not create `NSImage` out of the created \
+                                             `NSData` buffer".to_owned()));
+        }
+
+        unsafe {
+            let new_size = NSSize::new(ICON_WIDTH, ICON_HEIGHT);
+            let _: () = msg_send![nsimage, setSize:new_size];
+            tray_entry.button().setImage_(nsimage);
+            if let Ok(menu) = self.menu.get_mut(){
+                tray_entry.setMenu_(*menu);
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn add_menu_separator(&mut self, item_idx: u32) -> Result<(), Error> {
+        let item = unsafe { 
+            NSMenuItem::separatorItem(nil)
+        };
+        if item == nil {
+            return Err(Error::OsError("Could not create `NSMenuItem`."
+                                             .to_owned()));
+        }
+
+        unsafe {
+            if let Ok(menu) = self.menu.get_mut(){
+                NSMenu::addItem_(*menu, item);
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn add_menu_entry(&mut self, item_idx: u32, item_name: &str, callback: Callback) -> Result<(), Error> {
+        let blank_key = unsafe { NSString::alloc(nil).init_str("") };
+        if blank_key == nil {
+            return Err(Error::OsError("Could not create blank `NSString`."
+                                             .to_owned()));
+        }
+
+        let title = unsafe { NSString::alloc(nil).init_str(item_name) };
+        if title == nil {
+            return Err(Error::OsError("Could not create `NSString` from the item name."
+                                             .to_owned()));
+        }
+
+        let action = sel!(call);
+
+        let item = unsafe { 
+            NSMenuItem::alloc(nil)
+                .initWithTitle_action_keyEquivalent_(title, action, blank_key)
+        };
+        if item == nil {
+            return Err(Error::OsError("Could not create `NSMenuItem`."
+                                             .to_owned()));
+        }
+
+        unsafe {
+            if let Some(app) = &self.systray_application {
+                let _ : () = msg_send![item, setTarget: CocoaCallback::from(app.clone(), callback)];
+            }
+            if let Ok(menu) = self.menu.get_mut(){
+                NSMenu::addItem_(*menu, item);
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn shutdown(&self) -> Result<(), Error> {
+        Ok(())
+    }
 }
+
+unsafe impl Send for Window {}
```

Comment:
This does not seem right. `Window` contains an `Rc` which is clearly `!Send`, so just declaring `Window` to be `Send` is not sound.

Reply:
Removed in 3a00d0d